pycontrails/datalib/ecmwf/model_level_era5.py

```diff
@@ -0,0 +1,501 @@
+
+    Notes
+    -----
+    This function depends on `metview <https://metview.readthedocs.io/en/latest/python.html>`_
+    python bindings and binaries.
+
+    The lifetime of the metview import must last until processed datasets are cached
+    to avoid premature deletion of metview temporary files.
+    """
+    try:
+        import metview as mv
+    except ModuleNotFoundError as exc:
+        dependencies.raise_module_not_found_error(
+            "model_level.grib_to_dataset function",
+            package_name="metview",
+            module_not_found_error=exc,
+            extra="See https://metview.readthedocs.io/en/latest/install.html for instructions.",
+        )
+    except ImportError as exc:
+        msg = "Failed to import metview"
+        raise ImportError(msg) from exc
+
+    if era5.cachestore is None:
+        msg = "Cachestore is required to download and cache data"
+        raise ValueError(msg)
+
+    stack = contextlib.ExitStack()
+    request = era5.mars_request(times)
+
+    if not era5.cache_grib:
+        target = stack.enter_context(temp.temp_file())
+    else:
+        request_str = ";".join(f"{p}:{request[p]}" for p in sorted(request.keys()))
+        name = hashlib.md5(request_str.encode()).hexdigest()
+        target = era5.cachestore.path(f"era5ml-{name}.grib")
+
+    with stack:
+        if not era5.cache_grib or not era5.cachestore.exists(target):
+            if not hasattr(era5, "cds"):
+                era5._set_cds()
+            era5.cds.retrieve("reanalysis-era5-complete", request, target)
+
+        # Read contents of GRIB file as metview Fieldset
+        LOG.debug("Opening GRIB file")
+        fs_ml = mv.read(target)
+
+        # reduce memory overhead by cacheing one timestep at a time
+        for time in times:
+            fs_pl = mv.Fieldset()
+            dimensions = era5.ensemble_members if era5.ensemble_members else [-1]
+            for ens in dimensions:
+                date = time.strftime("%Y%m%d")
+                t = time.strftime("%H%M")
+                selection = dict(date=date, time=t)
+                if ens >= 0:
+                    selection |= dict(number=str(ens))
+
+                lnsp = fs_ml.select(shortName="lnsp", **selection)
+                for var in era5.variables:
+                    LOG.debug(
+                        f"Converting {var.short_name} at {t}"
+                        + (f" (ensemble member {ens})" if ens else "")
+                    )
+                    f_ml = fs_ml.select(shortName=var.short_name, **selection)
+                    f_pl = mv.mvl_ml2hPa(lnsp, f_ml, era5.pressure_levels)
+                    fs_pl = mv.merge(fs_pl, f_pl)
+
+            # Create, validate, and cache dataset
+            ds = fs_pl.to_dataset()
+            ds = ds.rename(isobaricInhPa="level").expand_dims("time")
+            ds.attrs["pycontrails_version"] = pycontrails.__version__
+            era5.cache_dataset(ds)
+
+
+class CDSCredentialsNotFound(Exception):
```

Comment:
Could this just be imported from the other `era5` module?

Reply:
Yes... will fix that

Reply:
ff29b019df9b83a4d08354566e3842d99c2ed3a4